### PR TITLE
Implement declarative config for compatibility with setuptools declarative config

### LINKED ISSUE #181: setuptools_scm does not work with setup.cfg
I was debuging  the behaviour of `setuptools_scm` because I received exception from function `version_keyword`
``` 
File "/home/oglop/.virtualenvs/b/lib/python3.6/site-packages/setuptools_scm/integration.py", line 22, in version_keyword
    dist.metadata.version = get_version(**value)
```

in `setuptools` file `config.py`, all options go through parser except additional options like `use_scm_version`, I guess this should be handled by `setuptools_scm`
*file* : `setuptools/config.py`
```
class ConfigOptionsHandler(ConfigHandler):

    section_prefix = 'options'

    @property
    def parsers(self):
        """Metadata item name to parser function mapping."""
        parse_list = self._parse_list
        parse_list_semicolon = partial(self._parse_list, separator=';')
        parse_bool = self._parse_bool
        parse_dict = self._parse_dict

        return {
            'zip_safe': parse_bool,
            'use_2to3': parse_bool,
            'include_package_data': parse_bool,
            'package_dir': parse_dict,
            'use_2to3_fixers': parse_list,
            'use_2to3_exclude_fixers': parse_list,
            'convert_2to3_doctests': parse_list,
            'scripts': parse_list,
            'eager_resources': parse_list,
            'dependency_links': parse_list,
            'namespace_packages': parse_list,
            'install_requires': parse_list_semicolon,
            'setup_requires': parse_list_semicolon,
            'tests_require': parse_list_semicolon,
            'packages': self._parse_packages,
            'entry_points': self._parse_file,
            'py_modules': parse_list,
        }
```

in `setuptools_scm`
 something goes wrong here:
```
[distutils.setup_keywords]
        use_scm_version = setuptools_scm.integration:version_keyword
```
because function `version_keyword` receives parameter `value` as `str` but that function expects `bool` - because the value did not go through the *parsing process*

what I do not know is how that `version_keyword` is called

### PR DESCRIPTION
**Pull request by jaraco**
Fixes #181 

In irc.freenode.net#pylib, Ronny asked about supporting setuptools_scm on setuptools in a declarative config world. This change illustrates how I'd propose for that to happen. Setuptools would expose a new hook, something like

https://gist.github.com/b45998781be03918a21f17da996ee687

And then setuptools_scm would provide a hook into that such as the one (building on all of the existing work already done).

### CODE REVIEW

**Comment by jaraco** [comment]
Still to do:

- [x] Settle on the right hook/API and release changes in setuptools.
- [x] Decide if setuptools_scm needs to support both `setup()` invocation and this declarative mode.
- [x] If so, implement compatibility protocol.
- [x] Add tests for this new functionality.
- [x] Implement the stubbed config file reader.
- [x] Update docs to describe how to use this preferred functionality.

**Review by RonnyPfannschmidt — COMMENTED** [review]
I like this,
the setup keyword can actually stay relatively easy with this,

Could setuptools provide a helper to get the settings section from pyproject.toml

Thanks!

**Comment by jaraco** [comment]
Unfortunately, setuptools itself doesn't yet have toml (or pyproject.toml) support... but it might be worthwhile for setuptools to own that functionality to avoid plugins needing to own the parsing.

How much harm is it for setuptools_scm to have a dependency (toml)?

**Comment by RonnyPfannschmidt** [comment]
its a pain for anyone using the legacy support as it will pull in toml via easy_install

so having a toml dependency outside of what setuptools does would be a big blow for backward compat

its possible to make it a extra dependency 
but its not clear to me how to make the entrypoint work properly then in all situation

**Comment by RonnyPfannschmidt** [comment]
@jaraco got thinking again, its easy to have a extra and a optional import, after all - i can sort out most of the pain i believe (bit it would still be much nicer if the toml data was readily available in some way)

**Comment by jaraco** [comment]
> its a pain for anyone using the legacy support as it will pull in toml via easy_install

Are there scenarios where requiring a pure-python source distribution wouldn't work as a simple dependency (as [toml installs under easy_install](https://gist.github.com/ed775dc0859eedb5c955b8ebf09079f4))? I'm not opposed to finding another solution, but I'd like to know what scenario we're seeking to solve.

> its easy to have a extra and a optional import

Oh, so `setuptools_scm[toml]>=version_with_toml` (or similar) would be the requirement specified by `setup_requires` or PEP 518 `build_requires`? That sounds feasible.

File: testing/test_config.py [hunk 1]
```diff
@@ -18,3 +18,10 @@ def test_tag_regex(tag, expected_version):
     match = config.tag_regex.match(tag)
     version = match.group("version")
     assert version == expected_version
+
+
+def test_config_from_pyproject(tmpdir):
+    fn = tmpdir / 'pyproject.toml'
+    fn.write_text('[setuptools_scm]\nenabled = true\n', encoding='utf-8')
```
Comment (RonnyPfannschmidt) [inline testing/test_config.py]: Needs to use the `tool.setuptools_scm` section

File: setup.py [hunk 2]
```diff
@@ -110,6 +113,7 @@ def parse(root):
         "Topic :: System :: Software Distribution",
         "Topic :: Utilities",
     ],
+    install_requires=["toml"],
```
Comment (RonnyPfannschmidt) [inline setup.py]: This should use an extra for now

**Review by RonnyPfannschmidt — COMMENTED** [review]
This is shaping up nicely, I like it

File: setup.py (hunk 2 above)
Comment (RonnyPfannschmidt) [reply setup.py]: I just noted the last comment which means just that

File: testing/test_config.py (hunk 1 above)
Comment (RonnyPfannschmidt) [inline testing/test_config.py]:
```suggestion
    fn.write_text('[tool.setuptools_scm]\nenabled = true\n', encoding='utf-8')
```

File: src/setuptools_scm/config.py [hunk 3]
```diff
@@ -105,3 +105,14 @@ def tag_regex(self):
...
+
+    @classmethod
+    def from_file(cls, name='pyproject.toml'):
+        """
+        Read Configuration from pyproject.toml (or similar)
+        """
+        with open(name) as strm:
+            defn = __import__('toml').load(strm)
+        config = cls()
+        vars(config).update(defn.get('setuptools_scm', {}))
```
Comment (RonnyPfannschmidt) [inline src/setuptools_scm/config.py]:
```suggestion
        vars(config).update(defn.get("tool",{}).get('setuptools_scm', {}))
```

Comment (RonnyPfannschmidt) [reply src/setuptools_scm/config.py]: this fun hack is definitively a little bit evil ^^

File: setup.py (hunk 2 above)
Comment (jaraco) [reply setup.py]: Done in 6fb77d9

File: testing/test_config.py [hunk 4]
```diff
@@ -18,3 +20,10 @@ def test_tag_regex(tag, expected_version):
     match = config.tag_regex.match(tag)
     version = match.group("version")
     assert version == expected_version
+
+
+def test_config_from_pyproject(tmpdir):
+    fn = tmpdir / "pyproject.toml"
+    fn.write_text("[tool.setuptools_scm]\nenabled = true\n", encoding="utf-8")
```
Comment (RonnyPfannschmidt) [inline testing/test_config.py]: i would like to suggest write_bytes + a bytes literal at this point

Comment (jaraco) [reply testing/test_config.py]: In my opinion, unicode literals are more friendly and the intended use here (a6ae830).

File: src/setuptools_scm/config.py [hunk 5]
```diff
@@ -49,6 +49,7 @@ class Configuration(object):
     parse = None
     _tag_regex = None
     _absolute_root = None
+    enabled = False
```
Comment (RonnyPfannschmidt) [inline src/setuptools_scm/config.py]: this needs to be handled in the legacy integration (setup_keyword)

since 
```
In [2]: toml.loads("[foo.bar]")                                                                                                                                                                                                                              
Out[2]: {'foo': {'bar': {}}}
```
i believe we should use the existence of the section as indicator for trigger rather than introducing a extra boolean

Comment (jaraco) [reply src/setuptools_scm/config.py]: > i believe we should use the existence of the section as indicator for trigger rather than introducing a extra boolean

Yes, I think I agree with that. I do wish to avoid booleans in general.

My reasoning behind having the boolean was that I was imagining some users might have `setuptools_scm` present and some configuration present for a number of reasons:

- they have a draft of the config but aren't yet ready to enable it
- they want to control some other aspect of setuptools_scm not related to versioning (file finding or maybe some future functionality)

Basically, I worry that if the section is `tool.setuptools_scm`, that's to general to imply "infer a version from scm metadata". If the section were `tool.setuptools_scm.infer_version`, I would agree whole-heartedly.

Comment (RonnyPfannschmidt) [reply src/setuptools_scm/config.py]: as the key feature of the package is version finding from scm metadata,
if people want to write up a draft and "hide it", they can use `tool._setuptools_scm`

in this case i believe that trying to account for non-existing or currently non-configurable features would constitute a premature optimization

should a new need arise, it needs to get a new name in the tool section

Comment (jaraco) [reply src/setuptools_scm/config.py]: Done in bf15ea9.

**Review by RonnyPfannschmidt — APPROVED** [review]
great work, what about the wip status

**Comment by jaraco** [comment]
I've got a few more commits to address the docs.

**Comment by RonnyPfannschmidt** [comment]
thanks for implementing and documenting this

with this fabulous work many are finally free of `setup.py` 

i plan to release mid week after including the ci/pre-commit fixups

File: README.rst [hunk 6]
```diff
@@ -13,9 +13,63 @@ It also handles file finders for the supported SCMs.
...
 
+``pyproject.toml`` usage
+------------------------
+
+The preferred way to configure ``setuptools_scm`` is to author
+settings in a ``tool.setuptools_scm`` section of ``pyproject.toml``.
+
+This feature requires Setuptools 42 or later, released in Nov, 2019.
+If your project needs to support build from sdist on older versions
+of Setuptools, you will need to also implement the ``setup.py usage``
```
Comment (webknjaz) [inline README.rst:24]: Looks like you placed the closing backticks too far...

**Comment by madig** [comment]
Just chiming in to say FUCK YEAH. Thanks! One less file lying around 👍

File: README.rst (hunk 6 above)
Comment (jaraco) [reply README.rst:24]: I wasn't quite sure how to handle that, given that the backticks are in the title. I'm happy to accept edits on it.

Comment (webknjaz) [reply README.rst:24]: Oh. I'm not sure either. Could probably try using a link syntax.
